## Scripts associated with 'Chromosome Evolution in Lepidoptera' manuscript

This repository contains all of the source code - that together with the supplementary files available on Zenodo -are required to reproduce all figures presented in 'Chromosome evolution in Lepidoptera' (Wright et al. 2023).
https://www.biorxiv.org/content/10.1101/2023.05.12.540473v1.abstract 

Each subdirectory contains a README with descriptions of the files it contains.

*Charlotte Wright, Blaxter lab, Tree of Life, Wellcome Sanger Institute*

The scripts were writted using R version 4.1.0 and Python version 3.9.9. 

The scripts require the following R packages: `phytools`, `dplyr`, `ggplot2`,  `treeio`, `aplot`, `ggtree`, `tidyverse`, `patchwork`, `scales`, `ggpubr`, `reshape2`, `ggrepel`.

1_genomes
Code and files related to the first results section ("Over two hundred chromosomally-complete lepidopteran genomes")

2_merian_elements
Code and files related to the second results section ("Thirty-two ancestral lepidopteran linkage groups

3_interchromosomal_rearrangements
Code and files related to the third results section ("Distribution of fusion and fission events across Lepidoptera")

4_highly_rearranged_genomes
Code and files related to the fourth results section ("Extensive rearrangemets in eight independent lineages")

5_sequence_contents
Code and files related to the fifth results section ("Understanding biases in chromosomal fusions in Lepidoptera")

6_fusion_consequences
Code and files related to the sixth results section ("Consequences of fusions")

repeat_annotation_files
FASTA files containing the repeat libraries for each species.

